frontend/src/utils/generate-warnings.ts

```diff
@@ -75,6 +90,473 @@ export function produceWarnings(schedule: Schedule): IWarning[] {
+    // shouldn't matter, summerII would just be empty.
+    if (year.isSummerFull) {
+      addCoursesToMap(year.summer1, taken);
+    } else {
+      addCoursesToMap(year.summer1, taken);
+      addCoursesToMap(year.summer2, taken);
+    }
+  }
+
+  let requirementGroups: string[] = major.requirementGroups;
+  let res: IRequirementGroupWarning[] = [];
+  for (const name of requirementGroups) {
+    let requirementGroup: IMajorRequirementGroup =
+      major.requirementGroupMap[name];
+    // todo: if req group is RangeSection or contains a ICourseRange process after all other requirement groups.
+    // this is because these reqs can be satisfied by a wide range of courses, and you don't want it using up
+    // a course that is need to satisfy a more constrained requirement.
+    let unsatisfiedRequirement:
+      | IRequirementGroupWarning
+      | undefined = produceUnsatifiedRequirement(
+      requirementGroup,
+      taken,
+      coursesUsed
+    );
+    if (unsatisfiedRequirement) {
+      res.push(unsatisfiedRequirement);
+    }
+  }
+  return res;
+}
+
+/**
+ * Produce an IUnsatisfiedRequirementGroup object if the requirementGroup hasn't been fully satisfied. Undefined otherwise.
+ * @param requirementGroup the requirement group to check.
+ * @param taken the Map of courses a student has on their schedule right now.
+ * @param coursesUsed a set of courses which already satisfy some requirement for the major.
+ */
+function produceUnsatifiedRequirement(
+  requirementGroup: IMajorRequirementGroup,
+  taken: Map<string, number>,
+  coursesUsed: Set<string>
+): IRequirementGroupWarning | undefined {
+  switch (requirementGroup.type) {
+    case "AND": {
+      return processAndSection(requirementGroup, taken, coursesUsed);
+    }
+    case "OR": {
+      return processOrSection(requirementGroup, taken, coursesUsed);
+    }
+    case "RANGE": {
+      return processRangeSection(requirementGroup, taken, coursesUsed);
+    }
+    default: {
+      return undefined;
+    }
+  }
+}
+
+/**
+ * Produce an IUnsatisfiedRequirementGroup object if the given ANDSection hasn't been fully satisfied. Undefined otherwise.
+ * @param requirementGroup the requirement group to check.
+ * @param taken the Map of courses a student has on their schedule right now.
+ * @param coursesUsed a set of courses which already satisfy some requirement for the major.
+ */
+function processAndSection(
+  requirementGroup: ANDSection,
+  taken: Map<string, number>,
+  coursesUsed: Set<string>
+): IRequirementGroupWarning | undefined {
+  let satisfied: CreditHourTracker = {
+    hoursCompleted: 0,
+  };
+  let messages: string[] = [];
+  for (const requirement of requirementGroup.requirements) {
+    let message: string | undefined = processRequirement(
+      requirement,
+      taken,
+      satisfied,
+      coursesUsed,
+      // zero because the credit hours don't matter in an AndSection--need to satisfy all requirements.
+      0
```

Comment:
Followup – I don't like this because it almost never gets used, and when it does, it's only in one case, for a minor speedup in efficiency (breaking out early) when processIOrCourse is called. However, I don't really see any point in taking it out, so I think it's fine to leave it in. However, do leave a nice juicy comment explaining what this parameter means/is used for in the processRequirement function, or somewhere else.

Reply:
This is actually not just for a minor speedup, it is also for correctness of the validation. You must break out early to ensure that you don't use up courses that might be candidates to satisfy other requirement groups.